database/json_file_storage.go

```diff
@@ -0,0 +1,175 @@
+package database
+
+import (
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+var _ RelayerDatabase = &JSONFileStorage{}
+
+type chainState map[string]string
+
+// JSONFileStorage implements RelayerDatabase
+type JSONFileStorage struct {
+	// the directory where the json files are stored
+	dir string
+
+	// Each network has its own mutex
+	// The chainIDs used to index the JSONFileStorage are created at initialization
+	// and are not modified afterwards, so we don't need to lock the map itself.
+	mutexes map[ids.ID]*sync.RWMutex
+	logger  logging.Logger
+}
+
+// NewJSONFileStorage creates a new JSONFileStorage instance
+func NewJSONFileStorage(logger logging.Logger, dir string, networks []ids.ID) (*JSONFileStorage, error) {
+	storage := &JSONFileStorage{
+		dir:     filepath.Clean(dir),
+		mutexes: make(map[ids.ID]*sync.RWMutex),
+		logger:  logger,
+	}
+
+	for _, network := range networks {
+		storage.mutexes[network] = &sync.RWMutex{}
+	}
+
+	_, err := os.Stat(dir)
+	if err == nil {
+		// dir already exist
+		// return the existing storage
+		return storage, nil
+	}
+
+	// 0755: The owner can read, write, execute.
+	// Everyone else can read and execute but not modify the file.
+	err = os.MkdirAll(dir, 0755)
+	if err != nil {
+		storage.logger.Error("failed to create directory",
+			zap.String("dir", dir),
+			zap.Error(err))
+		return nil, err
+	}
+
+	return storage, nil
+}
+
+// Get the latest chain state from the json database, and retrieve the value from the key
+func (s *JSONFileStorage) Get(chainID ids.ID, key []byte) ([]byte, error) {
+	currentState := make(chainState)
+	fileExists, err := s.read(chainID, &currentState)
+	if err != nil {
+		s.logger.Error(
+			"failed to read file",
+			zap.String("chainID", chainID.String()),
+			zap.Error(err),
+		)
+		return nil, err
+	}
+	if !fileExists {
+		return nil, errors.Errorf("file does not exist. chainID: %s", chainID)
+	}
+
+	if _, ok := currentState[string(key)]; !ok {
+		return nil, errors.Errorf("key does not exist. chainID: %s key: %s", chainID, key)
+	}
+
+	return []byte(currentState[string(key)]), nil
+}
+
+// Put the value into the json database. Read the current chain state and overwrite the key, if it exists
+func (s *JSONFileStorage) Put(chainID ids.ID, key []byte, value []byte) error {
+	currentState := make(chainState)
+	_, err := s.read(chainID, &currentState)
+	if err != nil {
+		s.logger.Error(
+			"failed to read file",
+			zap.String("chainID", chainID.String()),
+			zap.Error(err),
+		)
+		return err
+	}
+
+	currentState[string(key)] = string(value)
+
+	return s.write(chainID, currentState)
```

Comment:
Not sure if this is a problem in a use case, but we are not performing an atomic update in between the read and write since we are releasing the lock in between. Concurrent Puts can result in lost updates where one or more writes are lost.

If this can be a problem, then we should grab the lock before the read and release it after the write. We can do that by doing one of the following: move the locking out of read and write (and perform them in the functions that call them), use recursive locks (I don't think they are supported by default in golang), or add a flag to read/write to tell them that a lock is not required for this call.

Reply:
In our use case, only under extreme circumstances would there be concurrent writes to the same JSON file in such a way that could cause issues. In order for that to happen, we'd have to have logs be [processed](https://github.com/ava-labs/awm-relayer/blob/16af742920ee732323bf3c1430f86d44538a7951/vms/evm/subscriber.go#L123) in such a way that subsequent blocks would overlap in the DB write. Even if that were to occur, the worst that would happen would be a log being processed multiple times, which wouldn't be an issue for correctness. https://github.com/ava-labs/awm-relayer/pull/23 would guard against this as well.

That all being said, having the writes be atomic is a good idea, that eliminates this class of bugs altogether, so I've moved the locking behavior into `Put`/`Get` as you suggested.